library/src/commonMain/kotlin/com/badoo/reaktive/observable/CombineLatest.kt

```diff
@@ -2,19 +2,62 @@ package com.badoo.reaktive.observable
 
 import com.badoo.reaktive.disposable.CompositeDisposable
 import com.badoo.reaktive.disposable.Disposable
-import com.badoo.reaktive.utils.lock.newLock
-import com.badoo.reaktive.utils.lock.synchronized
+import com.badoo.reaktive.utils.serializer.serializer
+
+private val dummyValue = Any()
 
 fun <T, R> Collection<Observable<T>>.combineLatest(mapper: (List<T>) -> R): Observable<R> =
     observableByEmitter { emitter ->
         val disposables = CompositeDisposable()
         emitter.setDisposable(disposables)
-        val lock = newLock()
-        val dummy = Any()
-        val values = MutableList<Any?>(size) { dummy }
+        val values = MutableList<Any?>(size) { dummyValue }
         var readyValues: List<T>? = null
         var activeSourceCount = size
 
+        val serializer =
+            serializer<CombineLatestEvent<T>> { event ->
+                when (event) {
+                    is CombineLatestEvent.OnNext -> {
+                        values[event.index] = event.value
+
+                        if ((readyValues == null) && values.none { it === dummyValue }) {
+                            @Suppress("UNCHECKED_CAST")
+                            readyValues = values as List<T>
+                        }
+
+                        readyValues
+                            ?.toList()
+                            ?.let {
+                                try {
+                                    mapper(it)
+                                } catch (e: Throwable) {
+                                    emitter.onError(e)
+                                    return@serializer false
+                                }
+                            }
+                            ?.also(emitter::onNext)
+
+                        true
+                    }
+
+                    is CombineLatestEvent.OnComplete -> {
+                        activeSourceCount--
+
+                        val allCompleted = (activeSourceCount == 0) || (values[event.index] === dummyValue)
```

Comment:
(values[event.index] === dummyValue)
From what I understand if it hasn't emitted a value and it completes empty, then we must complete the entire thing. 
Not obvious, worth adding a quick comment.